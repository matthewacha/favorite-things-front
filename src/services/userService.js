import Api from './api';

export default {
  async postUser(params) {
    const data = await Api().post('/v1/user/', params);
    return data;
  },
};
