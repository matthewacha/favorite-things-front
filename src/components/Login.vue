<template>
  <div class="login">
    <div class="login-frame">
      <div class="login-frame-title">
        <div>Login to follow your favorite things</div>
      </div>
      <div class="login-frame-body">
        <v-form ref="form" id="form-name" v-model="valid" lazy-validation>
          <v-text-field id="login-input" v-model="name" :rules="nameRules" label="Name" :counter="10" required></v-text-field>

          <v-btn :disabled="!valid" color="success" @click="validate" >login</v-btn>
        </v-form>
      </div>
    </div>
  </div>
</template>

<script>
import userService from '@/services/userService';
import router from '@/router';

export default {
  name: 'Login',
  props: {},
  data: () => ({
    valid: true,
    name: '',
    nameRules: [
      v => !!v || 'Name is required',
      v => (v && v.length <= 20) || 'Name must be less than 20 characters',
    ],
  }),
  methods: {
    async validate() {
      if (this.$refs.form.validate()) {
        const data = {
          name: this.$refs.form.$el[0].value,
        };
        try {
          const response = await userService.postUser(data);
          this.$store.dispatch('user', response.data);
          localStorage.setItem('userObject', JSON.stringify(response.data));
          router.push('/favorites');
        } catch (e) {
          this.$store.dispatch('userError', e.message);
        }
      }
    },
  },
};
</script>

<!-- Add "scoped" attribute to limit CSS to this component only -->
<style scoped lang="scss">
h3 {
  margin: 40px 0 0;
}
ul {
  list-style-type: none;
  padding: 0;
}
li {
  display: inline-block;
  margin: 0 10px;
}
a {
  color: #42b983;
}
.login {
  height: inherit;
  align-content: center;
  width: 100%;
  text-align: center;
  background-image: url("../assets/funFront.jpg");
  .login-frame {
    height: 40%;
    width: 40%;
    position: absolute;
    left: 50%;
    top: 50%;
    transform: translate(-50%, -50%);
    border-radius: 4px;
    background-color: #2c3e5012;
    &-title {
      height: 20%;
      color: #fffcdb;
      font-weight: 500;
      font-size: 2em;
      padding-top: 40px;
    }
    &-body {
      padding: 10 !important;
      #form-name {
        margin: 20px;
        margin-top: 60px;
      }
    }
  }
}
.theme--light.v-btn:not(.v-btn--icon):not(.v-btn--flat) {
  background-color: #2b3e50 !important;
}
</style>
