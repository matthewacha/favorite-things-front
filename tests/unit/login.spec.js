import Vue from 'vue';
import { createLocalVue, shallowMount, mount } from '@vue/test-utils';
import sinon from 'sinon';
import Vuetify from 'vuetify';
import { expect } from 'chai';
import Vuex from 'vuex';
import userService from '../../src/services/userService';
import Login from '../../src/components/Login.vue';


Vue.use(Vuex);
Vue.use(Vuetify);

const userObject = {
  user: {
    id: 1,
  },
};
const localStorage = {
  getItem: user => JSON.stringify(userObject),
  setItem: (user) => {},
};
global.localStorage = localStorage;


describe('Login.vue', () => {
  it('renders login page properly', () => {
    const localVue = createLocalVue()
    localVue.use(Vuetify)
    const wrapper = shallowMount(Login, { sync: false });
    expect(wrapper.html()).to.contain('form');
  });
});

describe('test login functionality', () => {
  let actions;
  let store;
  const localVue = createLocalVue();
  localVue.use(Vuetify);

  beforeEach(() => {
    actions = {
      user: sinon.stub(),
    };
    store = new Vuex.Store({
      state: {},
      actions,
    });
  });

  it('redirects when login successful', () => {
    const wrapper = mount(Login, { sync: false, store });
    Object.defineProperty(userService, 'postUser', {
      value: () => Promise.resolve({
        data: {
          message: 'user exists',
          success: true,
          user: { name: 'ease-e', id: 36 },
        },
      }),
    });
    const spy = sinon.spy(userService, 'postUser');
    wrapper.vm.$refs.form = {
      $el: [{ value: 'ease-e' }],
    };
    wrapper.vm.$refs.form.validate = () => true;
    const submitForm = wrapper.find('button');
    submitForm.trigger('click');
    expect(spy.called).to.be.true;
  });

  it('it throws notification when error emitted', () => {
    const wrapper = mount(Login, { store });
    Object.defineProperty(userService, 'postUser', {
      value: () => Promise.reject(new Error('Network Error')),
    });
    const spy = sinon.spy(userService, 'postUser');
    wrapper.vm.$refs.form = {
      $el: [{ value: 'ease-e' }],
    };
    wrapper.vm.$refs.form.validate = () => true;

    const submitForm = wrapper.find('button');
    submitForm.trigger('click');
    expect(spy.called).to.be.true;
  });
});
