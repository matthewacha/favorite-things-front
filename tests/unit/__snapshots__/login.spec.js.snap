// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login.vue renders properly 1`] = `
<div
  class="login"
>
  <div
    class="login-frame"
  >
    <div
      class="login-frame-title"
    >
      <div>
        Login to follow your favorite things
      </div>
    </div>
     
    <div
      class="login-frame-body"
    >
      <v-form
        id="form-name"
        lazy-validation=""
      >
        <v-text-field
          counter="10"
          label="Name"
          required=""
          rules="v => {
          /* istanbul ignore next */cov_1fz8kj13j9.f[3]++;
          cov_1fz8kj13j9.s[8]++;
          return (/* istanbul ignore next */(cov_1fz8kj13j9.b[2][0]++, !!v) || /* istanbul ignore next */(cov_1fz8kj13j9.b[2][1]++, \\"Name is required\\")
          );
        },v => {
          /* istanbul ignore next */cov_1fz8kj13j9.f[4]++;
          cov_1fz8kj13j9.s[9]++;
          return (/* istanbul ignore next */(cov_1fz8kj13j9.b[3][0]++, v) && /* istanbul ignore next */(cov_1fz8kj13j9.b[3][1]++, v.length <= 20) || /* istanbul ignore next */(cov_1fz8kj13j9.b[3][2]++, \\"Name must be less than 20 characters\\")
          );
        }"
        />
         
        <v-btn
          color="success"
        >
          login
        </v-btn>
      </v-form>
    </div>
  </div>
</div>
`;
